### PR TITLE
feat: Add YaRN rope scaling support on Magatron-Bridge

### LINKED ISSUE #1678: Can we apply yarn to Nemo-RL
Relating to https://github.com/NVIDIA-NeMo/RL/pull/957

### PR DESCRIPTION
**Pull request by RayenTian**
# Summary
This PR adds YaRN long-context support for Qwen3-0.6B Megatron workflows and introduces matching SFT/GRPO recipes and test coverage.

# What changed
- Added new 1n8g Megatron YaRN 64k recipes for:
  - SFT: `sft-qwen3-0.6B-1n8g-megatron-yarn-64k`
  - GRPO: `grpo-qwen3-0.6B-1n8g-megatron-yarn-64k`
- Added matching test suite scripts for both recipes.
- Added `NemotronCascade2SFTMathDataset` and registered it in the response dataset registry.
- Updated Megatron setup to:
  - validate that YaRN `rope_scaling` fields are fully specified
  - force HF -> Megatron re-import when `hf_config_overrides` are present or when `force_reimport_model=true`
- Updated GRPO setup so vLLM generation inherits the policy HF overrides, ensuring generation uses the same YaRN rope settings as training.
- Added an Automodel guardrail to reject YaRN configs on the DTensor/Automodel backend, since YaRN is currently supported through the Megatron backend only.
- Documented the `force_reimport_model` option in the training backend docs.

# Issues
closes https://github.com/NVIDIA-NeMo/RL/issues/1678

# Test 

## SFT
image

## GRPO
- Yarn is working only when the seqlen exceeds the original model seqlen. Qwen2.5 and other models' original seqlens are about 128k or longer, which means we have to set seqlen to longer to enable yarn. And this'll lead to the long-tail issue being more notable, and further longer time_per_step.
    - For Qwen2.5-1.5B, seqlen=256k, some steps will take 1100+ seconds and 20 seconds for normal.
    - At the very beginning 20-30 steps, the `max_gen_tokens_per_sample` can reach 256K. This lead to long step time and high `token_mult_prob_error`. It will return to normal standard after few steps.
    - However, we cannot guarantee that setting max seqlen=256k will result in the rollout reaching 256k in every run.
    - Counterexample:  The maximum is only 24k
image

- Also tried Qwen3 models whose original context length is shorter (40k), while the original capabilities of Qwen3 models are too strong for the OpenMath dataset. 
    - At the very beginning, the reward is 1.
    - As the experiences go on, the validation/accuracy drops. Tried on shorter seqlen with rope, also meet this issue. So it's not because of yarn.
    - Qwen3 tend to generate longer rollout, which lead to longer step time.
    - Qwen3-0.6B results
image

- Also tried Qwen/Qwen2.5-Math-1.5B-Instruct, which has shorter original model seqlen(4k), so that we can get **faster** step time. 
    - This model has been trained in previous fine-tuning to discourage the generation of longer seqlen. Even thought set max_model_len to 8K, It only in several steps can generate long seqlen to 8k. 
    - It has strong initial mathematical capabilities, and the reward hardly increases.
    - Qwen/Qwen2.5-Math-1.5B-Instruct Results
image

  
All in all, finally choose qwen2.5-1.5B in the recipe. Its validation accuracy is increasing, and after the first 20-30 steps, it runs faster.

- In VLLM, seqlen exceed the original model length is invalid if we don't open yarn. So we can not compare seqlen=256k   results between yarn and rope. The result is the compare of `seqlen= 128k + rope` and `seqlen=256k + yarn`.

### Final GRPO results(Qwen2.5-1.5B)

image

For `max_gen_tokens_per_sample`, at the very beginning step, it can reach to 256k/128k
image

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [ ] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* ...

### CODE REVIEW

**Review by terrykong — COMMENTED** [review]
Minor review — 2 findings.

**[guideline: config-conventions]** `force_reimport_model` is used in `megatron/setup.py` and present in all exemplar YAMLs, but not declared in the `MegatronConfig` TypedDict (`nemo_rl/models/policy/__init__.py:199-251`). Consider adding it following the `force_overwrite_initial_ckpt` pattern on line 224:
```python
    # Force reimport of the HF model even if a converted checkpoint exists (default: False)
    force_reimport_model: NotRequired[bool]
```

File: nemo_rl/models/megatron/setup.py [hunk 1]
```diff
@@ -958,7 +978,10 @@ def handle_model_import(
     pt_checkpoint_exists: bool,
 ) -> None:
     """Handle HF model import if checkpoint doesn't exist."""
-    if pt_checkpoint_exists:
+    hf_config_overrides = config.get("hf_config_overrides", {}) or {}
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py]: [nit] `hf_config_overrides` is computed here and is already in scope for the `else` branch below, where the exact same expression is assigned again. The second assignment is redundant — consider removing it.

File: nemo_rl/models/megatron/setup.py [hunk 2]
```diff
@@ -958,7 +978,10 @@ def handle_model_import(
...
+    force_reimport_model = config["megatron_cfg"].get("force_reimport_model", False)
+
+    if pt_checkpoint_exists and hf_config_overrides == {} and not force_reimport_model:
```
Comment (terrykong) [inline nemo_rl/models/megatron/setup.py]: so IIUC, if the user now sets `hf_config_overrides` we'll always recreate the checkpoint on every restart of the model. i feel like we should do something more intelligent to avoid doing this all the time (maybe check against json seralized hf_config_overrides? 

otherwise seems like we'll always get a cache miss and do this over and over (larger models it can be more costly)

File: docs/design-docs/training-backends.md [hunk 3]
```diff
@@ -75,4 +75,16 @@ export HF_HOME="/shared/nfs/huggingface"
...
+- **Sufficient space**: Ensure adequate disk space for the converted model checkpoints.
+
+### Force Reimport ###
+
+By default, NeMo RL skips the HF → Megatron conversion if a converted checkpoint already exists at the target path. If you need to force a fresh conversion (e.g., after updating megatron-bridge or changing `hf_config_overrides`), set the following option in your config:
+
+```yaml
+policy:
+  megatron_cfg:
+    force_reimport_model: True  # Default: False
```
Comment (terrykong) [inline docs/design-docs/training-backends.md]: btw great addition! can we maybe rename this to `force_reconvert_from_hf` to be more explicit on what it does?

also, did you get a chance to test this? does it delete/overwrite all the previous files?

cc @yfw

File: nemo_rl/algorithms/grpo.py [hunk 4]
```diff
@@ -680,7 +680,7 @@ def initialize_generation_with_policy(
             )
 
         ## make vllm hf overrides match the training policy
-        generation_config["vllm_cfg"]["hf_overrides"] = policy_config.get(
+        generation_config["vllm_kwargs"]["hf_overrides"] = policy_config.get(
```
Comment (terrykong) [inline nemo_rl/algorithms/grpo.py:683]: so this change makes sense since this is meant for passthru, can you also clean this up here https://github.com/NVIDIA-NeMo/RL/blob/79269afc0/nemo_rl/models/generation/vllm/vllm_worker.py#L380-L382

the overwrite from vllm_cfg seems like a silent "gotcha". 

also distillation still uses vllm_cfg, we should unify https://github.com/NVIDIA-NeMo/RL/blob/7bd853af3/nemo_rl/algorithms/distillation.py#L413-L417

Comment (RayenTian) [reply nemo_rl/algorithms/grpo.py:683]: Done.

File: docs/design-docs/training-backends.md (hunk 3 above)
Comment (RayenTian) [reply docs/design-docs/training-backends.md]: Done changing name to `force_reconvert_from_hf`. Add two unit test to check if it is reconvert.

File: nemo_rl/models/megatron/setup.py (hunk 1 above)
Comment (RayenTian) [reply nemo_rl/models/megatron/setup.py]: Revert here since add hf_overrides hash.

File: nemo_rl/models/megatron/setup.py (hunk 2 above)
Comment (RayenTian) [reply nemo_rl/models/megatron/setup.py]: Add hf_overrides hash check to solve re-convert for same hf_overrides config

**Comment by RayenTian** [comment]
/ok to test [7140966](https://github.com/NVIDIA-NeMo/RL/pull/2188/commits/7140966e8630215de5dd6efe1a1df4e7943fff22)

File: examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+  checkpoint_dir: results/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k
+  save_period: 20
+policy:
+  train_global_batch_size: 32
+  train_micro_batch_size: 1
+  max_total_sequence_length: 262144
+  make_sequence_length_divisible_by: 64
+  megatron_cfg:
+    context_parallel_size: 32
+    force_reconvert_from_hf: true
```
Comment (yuki-97) [inline examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml]: nit: since we have hf hash, default false should be better? also for another yaml.

File: nemo_rl/data/datasets/response_datasets/nemotron_cascade2_sft.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    HuggingFace.  Each example already contains a ``messages`` field in
+    OpenAI chat format (system / user / assistant turns), so no heavy
+    reformatting is needed.
+
+    Args:
+        split: HuggingFace dataset split to load, default is "train"
+        split_validation_size: Fraction of data held out for validation when
+            no dedicated validation split exists, default is 0.05
+        seed: Random seed used when creating the train/validation split,
+            default is 42
```
Comment (yuki-97) [inline nemo_rl/data/datasets/response_datasets/nemotron_cascade2_sft.py]: nit
```suggestion
        seed: Random seed used when shuffling before selecting max_samples and
            when creating the train/validation split, default is 42
```

File: examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml (hunk 5 above)
Comment (yuki-97) [inline examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml]: what about adding it to `MegatronConfig` in `nemo_rl/models/policy/__init__.py`?

**Review by yuki-97 — COMMENTED** [review]
thanks Rayen, lgtm. left minor comments.

File: examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml (hunk 5 above)
Comment (RayenTian) [reply examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml]: Done.

File: nemo_rl/data/datasets/response_datasets/nemotron_cascade2_sft.py (hunk 6 above)
Comment (RayenTian) [reply nemo_rl/data/datasets/response_datasets/nemotron_cascade2_sft.py]: Done.

File: examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml (hunk 5 above)
Comment (RayenTian) [reply examples/configs/recipes/llm/grpo-qwen2.5-1.5B-4n8g-megatron-yarn-256k.yaml]: There's an attribute `force_overwrite_initial_ckpt` we defined it but never use it, change it into `force_reconvert_from_hf`

**Review by yuki-97 — APPROVED** [review]
